kvstreamer: support InOrder mode

yuzefovich opened the pull request.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75574)
<!-- Reviewable:end -->

yuzefovich committed:
WIP on InOrder mode

yuzefovich committed:
kvstreamer: create Results for empty Get responses

We recently merged a change to skip Get responses that came back empty
(because there is no correspoding key found) in the `Streamer` which
cleaned up some code. However, for the upcoming work on the `InOrder`
mode the `Streamer` needs to know about such responses. We could
introduce a separate tracking mechanism for empty Get responses and
continue not emitting them, but that seems like unnecessary complication
to the code, so this commit instead restores the creation of Results for
empty Get responses.

Release note: None

yuzefovich committed:
sql: only create a LeafTxn for local flows if Streamer is enabled

Previously, during the flow setup stage we would always choose to create
a LeafTxn for local flows if there is a processor that might use
a Streamer API. This was the case even when the Streamer is disabled by
a cluster setting. Such behavior is a regression (since leaf txns don't
have a transparent span refresh mechanism and read spans have to be
collected with the metadata at the end of the execution), so this commit
fixes things by only using a LeafTxn if the Streamer is actually
enabled. Now, the users of the Streamer API are expected to check that
they do have a LeafTxn and only use the Streamer if so.

Additionally, this commit correctly populates `HasConcurrency` field for
the local flows even when the query might be distributed. The fix is
needed since the physical planning decision to distribute the query or
not is finalized after we decide which txn to use for the flow, which is
too late. Thus, whenever the Streamer is enabled and we find an index
join in the plan, we will make sure to use the LeafTxn (this was already
the case if the plan ended up being distributed).

Release note: None
